File: javaBatchSample/src/main/jruby/lib/chunk/file/default_line_mapper.rb
# -*- encoding: utf-8 -*-

# TODO なぜか File.dirname(__FILE__) の戻り値がカレント (＝javaBatchSampleのトップディレクトリ)になる。
# 1. chdirしてみる。
# 2. java側で Ruby.reinitialize内でカレントを操作してみる。
require File.expand_path(File.join(File.dirname(__FILE__), "/src/main/jruby/lib/common"))


java_import 'org.springframework.batch.item.file.LineMapper'
java_import 'org.springframework.beans.factory.InitializingBean'
class DefaultLineMapper

  def afterPropertiesSet
    @count = 0
  end

  def mapLine( line, line_number)
    @count += 1
    p "mapLine開始...#{@count}回目"
    ret = nil
    begin
      if (line )
        ret = LogLine.new(line)
      end
      #pp ret
      ret
    rescue => ex
      # 発生した例外がjava由来のものかruby由来のそれかをチェック。
      # 次のどちらのコードでも書けるので、使いやすい方で。
      #p "例外発生。#{ex.class.name}     #{ex.is_a?(Java::JavaLang::Throwable)}"
      #p "例外発生。#{ex.class.name}     #{ex.is_a?(java.lang.Throwable)}"
      
      # rubyで定義された例外をそのままjava例外のコンストラクタに渡してもダメ。ラップしてあげる必要がある。
      # TODO RaiseException だと必然的にRuntimeExceptionになってしまうので、もっといい方法がないか模索する。
      unless ( ex.is_a?(java.lang.Throwable) )
        ex = org.jruby.exceptions.RaiseException.new(ex)
      end
      
      raise org.springframework.batch.item.file.FlatFileParseException.new("Parsing error at line: #{line_number}, input=[#{line}]", ex, line, line_number)
    ensure
      p "mapLine終わり..."
    end
    
  end

end

p "これからnewします...."
DefaultLineMapper.new    # ←スクリプト最後のインスタンスを保持しておいて、必要の都度各メソッドを実行するイメージ

File: javaBatchSample/src/main/java/com/i3systems/momiken/jbatch/webserverlog/DefaultLineMapper.java
/**
 * Copyright (c) 2013 Oracle and/or its affiliates. All rights reserved.
 *
 * You may not modify, use, reproduce, or distribute this software except in
 * compliance with  the terms of the License at:
 * http://java.net/projects/javaeetutorial/pages/BerkeleyLicense
 */
package com.i3systems.momiken.jbatch.webserverlog;


import java.util.logging.Level;
import java.util.logging.Logger;

import org.springframework.batch.item.file.FlatFileParseException;
import org.springframework.batch.item.file.LineMapper;
import org.springframework.beans.factory.InitializingBean;





import com.i3systems.momiken.jbatch.webserverlog.items.LogLine;


public class DefaultLineMapper<T> implements LineMapper<LogLine>, InitializingBean {
    private static final Logger logger = Logger.getLogger("DefaultLineMapper");
    private int count;

	public void afterPropertiesSet() {
		count = 0;
	}
	
	public LogLine mapLine(String line, int lineNumber) throws Exception {
		
		count++;
        logger.log(Level.INFO, "DefaultLineMapper " + count + "回目");

		
		try{
			if (line != null) {
	            return new LogLine(line);
	        } else
	            return null;
		}
		catch(Exception ex){
			throw new FlatFileParseException("Parsing error at line: " + lineNumber + 
					", input=[" + line + "]", ex, line, lineNumber); 
		}
	}
}
